errors: protocol error refactor

muzarski opened the pull request:
ref: https://github.com/scylladb/scylla-rust-driver/issues/519

## QueryError::ProtocolError
In this PR we do a refactor of the `QueryError::ProtocolError` variant, which was represented simply as `&'static str`. In this PR, we create a new type `ProtocolError` which will represent all posible kinds of protocol errors. We simply replace all occurrences of `QueryError::ProtocolError` with a corresponding `ProtocolError` variant.

## QueryError::InvalidMessage
This variant is no longer use after the fixes. Previously, this variant was constructed for some protocol errors due to limitations of the `QueryError::ProtocolError` variant type (`&'static str`, when `String` was needed). After introducing `ProtocolError` type, this is no longer an issue. This variant can be safely removed.

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- ~[ ] I added relevant tests for new features and bug fixes.~
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- ~[ ] I have adjusted the documentation in `./docs/source/`.~
- [x] I added appropriate `Fixes:` annotations to PR description.

muzarski commented:
Rebased on main

wprzytula commented on scylla/src/transport/errors.rs:
```diff
@@ -355,6 +359,19 @@ pub enum KeyspaceStrategyProtocolError {
+    #[error("system_schema.types has invalid column type: {0}")]
+    SchemaTypesInvalidColumnType(FromRowError),
+
+    /// Cilcular UDT dependency detected.
+    #[error("Cilcular user defined types dependency detected - toposort is impossible!")]
```
Suggested rephrase: `Detected circular dependency between user defined types - toposort is impossible!`

wprzytula commented on scylla/src/transport/errors.rs:
```diff
@@ -355,6 +359,19 @@ pub enum KeyspaceStrategyProtocolError {
+
+    /// Cilcular UDT dependency detected.
+    #[error("Cilcular user defined types dependency detected - toposort is impossible!")]
+    CilcularTypeDependency,
```
ditto

wprzytula commented on scylla/src/statement/prepared_statement.rs:
```diff
@@ -237,13 +237,9 @@ impl PreparedStatement {
-                        QueryError::ProtocolError("No pk indexes - can't calculate token")
-                    }
-                })?;
+        let partition_key = self
+            .extract_partition_key(serialized_values)
+            .map_err(|_| ProtocolError::PartitionKeyExtraction)?;
```
That's risky that you no longer pattern match on the underlying error. This way, if someone extends `PartitionKeyExtractionError` with a new different variant, then there is a room for a bug - the new variant might be still converted to `ProtocolError::PartitionKeyExtraction` with a misleading description and the compiler won't complain.

wprzytula commented on scylla/src/transport/retry_policy.rs:
```diff
@@ -299,7 +299,7 @@ mod tests {
                 .to_owned(),
         )));
-        default_policy_assert_never_retries(QueryError::ProtocolError("test"));
+        default_policy_assert_never_retries(ProtocolError::RepreparedIdChanged.into());
```
Why such choice of `ProtocolError` variant?

muzarski replied:
No specific reason - we just want to test that DRP decides that we should not retry the query when protocol error occurred.

muzarski replied:
And this is variant which can be easily constructed (without additional arguments)

Lorak-mmk commented on scylla/src/transport/connection.rs:
```diff
@@ -1359,23 +1360,38 @@ impl Connection {
-                    return Err(QueryError::ProtocolError(
-                        "USE <keyspace_name> returned response with different keyspace name",
-                    ));
+                let expected_keyspace_name_lowercase = keyspace_name.as_str().to_lowercase();
+                let result_keyspace_name_lowercase = set_keyspace.keyspace_name.to_lowercase();
+
+                if result_keyspace_name_lowercase != expected_keyspace_name_lowercase {
+                    return Err(ProtocolError::UseKeyspace(
+                        UseKeyspaceProtocolError::KeyspaceNameMismatch {
+                            expected_keyspace_name_lowercase,
+                            result_keyspace_name_lowercase,
+                        },
+                    )
+                    .into());
```
Why do you want to use custom comparator when https://doc.rust-lang.org/std/primitive.str.html#method.eq_ignore_ascii_case exists?

Lorak-mmk commented on scylla/src/transport/topology.rs:
```diff
@@ -1698,25 +1705,19 @@ async fn query_table_partitioners(
-            let replication_factor: usize = usize::from_str(&rep_factor_str).map_err(|_| {
-                QueryError::ProtocolError("Could not parse replication factor as an integer")
-            })?;
+            let rep_factor_str: String = strategy_map
+                .remove("replication_factor")
+                .ok_or(KeyspaceStrategyProtocolError::MissingReplicationFactorForSimpleStrategy)?;
+
+            let replication_factor: usize = usize::from_str(&rep_factor_str)
+                .map_err(KeyspaceStrategyProtocolError::ReplicationFactorParseError)?;
```
Definitely. If there can be other things in this map, then we should know what are those things and filter them, and only them, out - not assume that everything can fail and ignore it.

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -66,6 +66,10 @@ pub enum QueryError {
+    /// Protocol error.
+    #[error("Protocol error: {0}")]
+    ProtocolErrorTyped(#[from] ProtocolError),
+
```
Thanks, I'll keep that in mind

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -261,6 +261,10 @@ pub enum ProtocolError {
     #[error("Schema version fetch protocol error: {0}")]
     SchemaVersionFetch(SingleRowTypedError),
 
+    /// A protocol error appeared during cluster metadata fetch.
+    #[error("Cluster metadata fetch protocol error: {0}")]
+    Metadata(#[from] MetadataProtocolError),
+
```
I'll investigate it.

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -301,6 +302,10 @@ pub enum MetadataProtocolError {
+
+    /// Bad keyspaces metadata.
+    #[error("Bad keyspaces metadata: {0}")]
+    Keyspaces(#[from] KeyspacesMetadataProtocolError),
```
Yes, you are right. I'll refactor it, so we have a separate type (`MetadataError`, as you mentioned) for all of the functions related to metadata fetch - including the public API, e.g. `Session::refresh_metadata` (which currently returns `QueryError`).

muzarski replied:
Hmm, now that I look at the code, `Session::refresh_metadata` is called in `Session::handle_auto_await_schema_agreement`. Auto schema agreement is performed in `Session::query/execute`. This is why we can't really remove metadata errors from `QueryError`. However, I still agree that this should not be included in `ProtocolError`, but rather directly in `Query/NewSessionError`.

avikivity commented on scylla/src/transport/topology.rs:
```diff
@@ -750,16 +751,18 @@ async fn query_metadata(
+        .into());
     }
 
     // At least one peer has to have some tokens
     if peers.iter().all(|peer| peer.tokens.is_empty()) {
```
ScyllaDB will continue to support vnodes for at least two years. But I expect the rust driver will continue to support vnodes forever, if we maintain Cassandra compatibility.

wprzytula replied:
I'd argue that this check is no longer correct/necessary, because it may bail out for operational clusters in the future.

avikivity replied:
We can choose to retain 1 token per node, just to avoid confusing drivers. But if the check can be safely removed, why not.

Lorak-mmk replied:
@muzarski Please investigate if this check is just hardening and can be removed or if other parts of the code depends on it.

muzarski replied:
It seems that default LBP relies on this assumption. When all of the nodes have empty token lists, the LBP returns an empty plan even for non-token-aware requests.

muzarski commented on scylla/src/transport/connection.rs:
```diff
@@ -1359,23 +1360,38 @@ impl Connection {
+        self.handle_use_keyspace_result(keyspace_name, query_response)
+    }
 
+    pub(crate) fn handle_use_keyspace_result(
+        &self,
+        keyspace_name: &VerifiedKeyspaceName,
+        query_response: QueryResponse,
+    ) -> Result<(), QueryError> {
```
Because of `Response::Error()` match branch - when server responds with CQL ERROR response. It is then converted to `QueryError::DbError`

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -261,6 +261,10 @@ pub enum ProtocolError {
+    /// A protocol error appeared during cluster metadata fetch.
+    #[error("Cluster metadata fetch protocol error: {0}")]
+    Metadata(#[from] MetadataProtocolError),
+
```
Since I think it makes sense to do it for this variant. However, IMO it's not true for variants such as `UnexpectedResponse(CqlResponseKind)` - here, `CqlResponseKind` is not an error type itself, it's there only to provide more context. The error itself is that we received an unexpected response - I don't think `From<CqlResponseKind>` impl makes sense here.

muzarski commented on scylla/src/transport/errors.rs:
```diff
@@ -225,6 +234,16 @@ pub enum NewSessionError {
 
+/// A protocol error.
+///
+/// It indicates an inconsistency between CQL protocol
+/// and server's behavior.
+/// In some cases, it could also represent a misuse
+/// of internal driver API - a driver bug.
```
Well, in some cases it's really hard to distinguish whose fault it is - look at the `NonfinishedPagingState` for example. It suggests that this MAY be a driver's internal API misuse.

muzarski commented:
v2: addressed review comments

Most important change is that I removed `Protocol` infix from errors related to metadata (e.g. `MetadataProtocolError` -> `MetadataError`).  I also moved `MetadaError` out of `ProtocolError`, directly under `[Query/NS]Error`.

Lorak-mmk commented on scylla/src/transport/errors.rs:
```diff
@@ -256,6 +256,15 @@ pub enum ProtocolError {
     /// A result with nonfinished paging state received for unpaged query.
     #[error("Unpaged query returned a non-empty paging state! This is a driver-side or server-side bug.")]
     NonfinishedPagingState,
+
+    /// Load balancing policy returned an empty plan - a driver bug!
+    #[error(
+        "Load balancing policy returned an empty plan.\
+        First thing to investigate should be the logic of custom LBP implementation.\
+        If you think that your LBP implementation is correct, or you make use of `DefaultPolicy`,\
+        then this is most probably a driver bug!"
+    )]
+    EmptyPlan,
```
How is empty plan related to protocol? I see no reason for it to be in `ProtocolError`.

wprzytula replied:
This is violation of a driver's invariant (driver's inner assumption), but it's not necessarily about the protocol (communication with server)

muzarski replied:
True, I've put it in `ProtocolError` without giving it a thought - just because previous version would return `QueryError::ProtocolError(...)`.

muzarski commented on scylla/src/transport/connection.rs:
```diff
@@ -1359,24 +1360,41 @@ impl Connection {
+        self.handle_use_keyspace_result(keyspace_name, query_response)
+    }
 
+    fn handle_use_keyspace_result(
+        &self,
+        keyspace_name: &VerifiedKeyspaceName,
+        query_response: QueryResponse,
+    ) -> Result<(), QueryError> {
```
> Another think: in recent force push you changed this method from pub(crate) to pub? Why?

Wojciech suggested to remove it: https://github.com/scylladb/scylla-rust-driver/pull/1080#discussion_r1777223293

wprzytula replied:
From what I see, it's changed from `pub(crate)` to `pub(self)` (implicit, because no visibility specifier is written).

muzarski replied:
> I'm not sure this is the best name for this method. It doesn't really handle anything - maybe `check_use_keyspace_result`?
> 
> Also it doesn't use self in the body - it can be a free function.

Removed `&self` parameter and renamed to `verify_use_keyspace_result`.

muzarski replied:
> From what I see, it's changed from `pub(crate)` to `pub(self)` (implicit, because no visibility specifier is written), not `pub` as @Lorak-mmk stated.

Oh, I misread the comment. That's right, we decreased the visibility

muzarski commented:
v2.1: addressed @Lorak-mmk comments
